program-runtime/src/loaded_programs.rs

```diff
@@ -720,35 +720,43 @@ impl LoadedPrograms {
     /// Extracts a subset of the programs relevant to a transaction batch
     /// and returns which program accounts the accounts DB needs to load.
     pub fn extract<S: WorkingSlot>(
-        &self,
+        &mut self,
         working_slot: &S,
         keys: impl Iterator<Item = (Pubkey, (LoadedProgramMatchCriteria, u64))>,
     ) -> ExtractedPrograms {
-        let environments = self.get_environments_for_epoch(working_slot.current_epoch());
+        let environments = self
+            .get_environments_for_epoch(working_slot.current_epoch())
+            .clone();
+        let loaded = Arc::new(Mutex::new(LoadedProgramsForTxBatch {
+            entries: HashMap::new(),
+            slot: working_slot.current_slot(),
+            environments: environments.clone(),
+        }));
         let mut missing = Vec::new();
-        let mut unloaded = Vec::new();
         let found = keys
```

Comment:
Not strictly related to this PR since it was already like this, but I think this
should be a loop not a filter_map.

The iterator is filtering `found`, and at the same time populating `missing`, setting some stats,
and now also populating `waiting_list`. Seems arbitrary and confusing that one is done through the
iterator and the rest imperatively. As a rule of thumb, I think iterators should be for one sequence
-> one sequence transformations.